src/emitter/batch_emitter.rs

```diff
@@ -0,0 +1,318 @@
+            // The currently running tokio tasks
+            let mut tokio_tasks: Vec<_> = Vec::new();
+
+            loop {
+                // `rx.recv().await` will not resolve until either a message is recieved,
+                // or the channel is closed and there are no more messages, in which case we exit the loop
+                let message = match rx.recv().await {
+                    Some(msg) => msg,
+                    None => break,
+                };
+
+                match message {
+                    EmitterMessage::Send(batch) => {
+                        // Clone HttpClient to be moved into the task
+                        let client = http_client.clone();
+
+                        // Spawn a new task to send the batch
+                        tokio_tasks.push(tokio::spawn(async move {
+                            // TODO: Change println to logging
+                            match Self::send_batch(batch, client).await {
+                                Ok(_) => println!("Batch sent"),
+                                Err(e) => println!("Error sending batch: {e}"),
+                            }
+                        }));
+                    }
+
+                    // On break, the emitter and runtime will be dropped
+                    //
+                    // Tokio will cancel any running tasks once the runtime is dropped, meaning any queued batches will be lost,
+                    // so we attempt to send any remaining batches before exiting
+                    EmitterMessage::Close => {
+                        for task in tokio_tasks {
+                            task.await.unwrap();
+                        }
+                        break;
+                    }
+                }
+
+                // Discard any completed tasks in the task list
+                tokio_tasks.retain(|t| !t.is_finished());
+            }
+        });
+    }
+}
+
+impl Drop for BatchEmitter {
+    fn drop(&mut self) {
+        // Get the join handle for the thread running the tokio runtime and wait for it to finish
+        //
+        // It's likely that the thread has already finished once the emitter loop has exited
+        if let Some(handle) = self.executor_handle.take() {
+            handle.join().unwrap();
+        }
+    }
+}
+
+impl Emitter for BatchEmitter {
+    /// Adds a payload to the event store
+    ///
+    /// This may also trigger sending a payload to the collector if the event store has enough events to fill a batch
+    fn add(&mut self, payload: PayloadBuilder) -> Result<(), Error> {
+        match self.event_store.lock() {
+            Ok(mut store) => match store.add(payload) {
+                Ok(_) => Ok(()),
+                Err(e) => Err(Error::EmitterError(e.to_string())),
+            },
+            Err(e) => Err(Error::EmitterError(e.to_string())),
+        }?;
+
+        let batch: Option<EventBatch> = {
+            let mut store_lock = self.event_store.lock().unwrap();
```

Comment:
Nit: Why do we unwrap the lock here? Isn't there a chance that it could panic? Also, could the call for `store.batch()` be incorporated in the lock above? Seems a bit wasteful to ask for lock twice.

Reply:
This has been removed now and the batch fetched in the previous lock block.